### PR TITLE
Specht modules for Ariki-Koike algebras

### PR DESCRIPTION
**Pull request by tscrim**
We provide an implementation of Specht modules for the Ariki-Koike algebra $H_{r,n}(q,u)$ over a field (more generally, a commutative ring with $q^k u_i - u_j$ for all applicable $i, j,k$ are invertible). To help ease the construction, we add an additional optional argument to the AK algebra constructor to have the base ring convert to the fraction field.

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: src/sage/algebras/hecke_algebras/ariki_koike_algebra.py [hunk 1]
```diff
@@ -411,6 +414,21 @@ def a_realization(self):
         """
         return self.LT()
 
+    def specht_module(self, la):
+        r"""
+        Return the Specht module of ``self`` corresponded to shape ``la``.
```
Comment (mkoeppe) [inline src/sage/algebras/hecke_algebras/ariki_koike_algebra.py]:
```suggestion
        Return the Specht module of ``self`` corresponding to shape ``la``.
```

File: src/sage/algebras/hecke_algebras/ariki_koike_algebra.py [hunk 2]
```diff
@@ -1183,6 +1219,18 @@ def __init__(self, algebra):
...
 
+        def _basis_to_word(self, t):
+            """
+            Return the basis element indexed by ``m`` to a word.
+            """
+            redword = []
+            for i, k in enumerate(t[0]):
+                if not k:
+                    continue
+                redword.extend(list(reversed(range(1, i+1))) + [0] * k)
```
Comment (mkoeppe) [inline src/sage/algebras/hecke_algebras/ariki_koike_algebra.py]:
```suggestion
                redword.extend(list(range(i, 0, -1)) + [0]*k)
```

File: src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+        cs = t.cells_containing(i+1)[0]
+        if len(ct) == 2: # it is of level 1 and a regular tableau
+            ct = (0,) + ct
+            cs = (0,) + cs
+
+        if ct[0] == cs[0] and ct[2] == cs[2]: # same column
+            return self.element_class(self, {t: -R.one()})
+
+        if ct[0] == cs[0] and ct[1] == cs[1]: # same row
+            return self.element_class(self, {t: self._q})
```
Comment (mkoeppe) [inline src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py]:
```suggestion
        if ct[0] == cs[0] and ct[2] == cs[2]:  # same column
            return self.element_class(self, {t: -R.one()})

        if ct[0] == cs[0] and ct[1] == cs[1]:  # same row
            return self.element_class(self, {t: self._q})
```

File: src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+        #   corresponding action of the permutation on t.
+        one = self.base_ring().one()
+        denom = res(cs) - res(ct)
+        coefft = (self._q - one) * res(cs) / denom
+        coeffs = (self._q * res(ct) - res(cs)) / denom
+        return self.element_class(self, {t: R(coefft), s: R(coeffs)})
+
+    def ariki_koike_algebra(self):
+        r"""
+        Return the Ariki-Koike algebra that ``self`` is a reprensetation of.
```
Comment (mkoeppe) [inline src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py]:
```suggestion
        Return the Ariki-Koike algebra that ``self`` is a representation of.
```

File: src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+                 + 3*u2/q*S[2,3|-|1,4,5/6] + u2/q*S[2,4|-|1,3,5/6]
+                sage: elt.L([3,3,3])
+                u2^3*S[1,2|-|3,4,5/6] + 2*u0^3*q^3*S[1,3|-|2,4,5/6]
+                + 3*u0^3*q^3*S[2,3|-|1,4,5/6] + u2^3*q^3*S[2,4|-|1,3,5/6]
+                sage: LT = AK.LT()
+                sage: elt.L([3,3,3]) == elt * (LT.L(3)^3)
+                True
+                sage: TableauTuples.options._reset()  # reset
+            """
+            if not self: # action on 0 is 0
```
Comment (mkoeppe) [inline src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py]:
```suggestion
            if not self:  # action on 0 is 0
```

File: src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+                ((u2*q-u2)/(-u0*q+u2))*S[2,4/8|-|1,5,7/3,6/9,10]
+                 + ((u0*q^2-u2)/(-u0*q+u2))*S[3,4/8|-|1,5,7/2,6/9,10]
+                sage: b.T(6)
+                ((-q)/(q+1))*S[2,4/8|-|1,5,6/3,7/9,10]
+                 + (q^2/(q+1))*S[2,4/8|-|1,5,7/3,6/9,10]
+                sage: b.T([2,1,2]) == b.T([1,2,1])
+                True
+                sage: b.T(9)
+                q*S[2,4/8|-|1,5,7/3,6/9,10]
+                sage: all(b.T([i,i]) == (q-1) * b.T(i) + q*b for i in range(1,10))
```
Comment (mkoeppe) [inline src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py]:
```suggestion
                sage: all(b.T([i,i]) == (q-1)*b.T(i) + q*b for i in range(1,10))
```

File: src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+
+    We will now describe the right action of the Ariki-Koike algebra,
+    but we first need to set some notation and definitions.
+    Let `t` be a standard tableau tuple of level `r` and size `n`.
+    Define the *residue* of `i` in `t` to be `r_t(i) = q^{c-r} u_k`,
+    where `i` is in cell `(r, c)` of the `k`-th tableau.
+
+    The action of `L_i` is given by `t \cdot L_i = r_T(i) t`. For `T_i`,
+    we need to consider the following cases. If `i, i+1` are in the same
+    row (resp. column), then `t \cdot T_i = q t` (resp. `t \cdot T_i = - t`).
```
Comment (mkoeppe) [inline src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py]:
```suggestion
    row (resp. column), then `t \cdot T_i = q t` (resp. `t \cdot T_i = -t`).
```

File: src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+
+        t \cdot T_i = \frac{(q - 1) r_t(i)}{r_s(i) - r_t(i)} t
+        + \frac{q r_t(i) - r_s(i)}{r_s(i) - r_t(i)} s.
+
+    Note that `r_s(i) = r_t(i+1)`.
+
+    Over a field of characteristic `0`, the set of Specht modules for all
+    partition tuples of level `r` and size `n` form the complete set
+    of irreducible modules for `H_{r,n}(q, u)` [AK1994]_. (The condition
+    on the base ring can be  weakened; see Theorem 3.2 of [Mathas2002]_.)
```
Comment (mkoeppe) [inline src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py]:
```suggestion
    on the base ring can be weakened; see Theorem 3.2 of [Mathas2002]_.)
```

File: src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py [hunk 9]
```diff
@@ -0,0 +1,473 @@
...
+            \lr{\phantom{x}}&\lr{\phantom{x}}\\\cline{1-2}
+            \lr{\phantom{x}}\\\cline{1-1}
+            \end{array}$}
+            }}_{\mathcal{H}_{3,8}(q)}
+        """
+        return "S^{{{}}}_{{{}}}".format(latex(self._shape), latex(self._AK))
+
+    def _test_representation(self, **options):
+        r"""
+        Test that the relations of the Ariki-Koike algebra are statisfied.
```
Comment (mkoeppe) [inline src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py]:
```suggestion
        Test that the relations of the Ariki-Koike algebra are satisfied.
```

File: src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py [hunk 10]
```diff
@@ -0,0 +1,473 @@
...
+
+        EXAMPLES::
+
+            sage: q = ZZ['q'].fraction_field().gen()
+            sage: AK = algebras.ArikiKoike(2, 4, q, [q^2+1,q-3], q.parent())
```
Comment (mkoeppe) [inline src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py]:
```suggestion
            sage: AK = algebras.ArikiKoike(2, 4, q, [q^2 + 1, q - 3], q.parent())
```

File: src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py [hunk 11]
```diff
@@ -0,0 +1,473 @@
...
+            u0*q*S[([[2, 4], [8]], [], [[1, 3, 7], [5, 6], [9, 10]])]
+            sage: S._L_on_basis(6, t)
+            u2*S[([[2, 4], [8]], [], [[1, 3, 7], [5, 6], [9, 10]])]
+            sage: S._L_on_basis(8, t)
+            (u0*q^-1)*S[([[2, 4], [8]], [], [[1, 3, 7], [5, 6], [9, 10]])]
+            sage: S._L_on_basis(9, t)
+            (u2*q^-2)*S[([[2, 4], [8]], [], [[1, 3, 7], [5, 6], [9, 10]])]
+        """
+        c = t.cells_containing(i)[0]
+        if len(c) == 2: # it is of level 1 and a regular tableau
```
Comment (mkoeppe) [inline src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py]:
```suggestion
        if len(c) == 2:  # it is of level 1 and a regular tableau
```

File: src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py [hunk 12]
```diff
@@ -0,0 +1,473 @@
...
+            c = (0,) + c
+        res = self._q**(c[2] - c[1]) * self._u[c[0]]
```
Comment (mkoeppe) [inline src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py]:
```suggestion
        res = self._q**(c[2]-c[1]) * self._u[c[0]]
```

File: src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py [hunk 13]
```diff
@@ -0,0 +1,473 @@
...
+
+        if ct[0] == cs[0] and ct[1] == cs[1]: # same row
+            return self.element_class(self, {t: self._q})
+
+        # result is standard
+        s = t.symmetric_group_action_on_entries(self._Pn.simple_reflection(i))
+        assert s.parent() is t.parent()
+
+        def res(cell):
+            return self._q**(cell[2] - cell[1]) * self._u[cell[0]]
```
Comment (mkoeppe) [inline src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py]:
```suggestion
            return self._q**(cell[2]-cell[1]) * self._u[cell[0]]
```

File: src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py [hunk 14]
```diff
@@ -0,0 +1,473 @@
...
+
+        # Note that the residue of i in t is given by the cell c
+        #   and of i in s corresponds to cell cp because the
+        #   corresponding action of the permutation on t.
+        one = self.base_ring().one()
+        denom = res(cs) - res(ct)
+        coefft = (self._q - one) * res(cs) / denom
+        coeffs = (self._q * res(ct) - res(cs)) / denom
```
Comment (mkoeppe) [inline src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py:315]:
```suggestion
        coeffs = (self._q*res(ct) - res(cs)) / denom
```

File: src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py [hunk 15]
```diff
@@ -0,0 +1,473 @@
...
+                ...
+                TypeError: unsupported operand parent(s) for *: 'Ariki-Koike algebra ... over Integer Ring'
+                sage: TableauTuples.options._reset()
+
+            TESTS::
+
+                sage: AK = algebras.ArikiKoike(2, 4, use_fraction_field=True)
+                sage: LT = AK.LT()
+                sage: T = AK.T()
+                sage: S = AK.specht_module([[1],[2,1]])
```
Comment (mkoeppe) [inline src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py]:
```suggestion
                sage: S = AK.specht_module([[1], [2,1]])
```

**Review by mkoeppe — COMMENTED** [review]
I cannot review the mathematical details, but the implementation looks clean. This PR has sat long enough, so let's merge it. I've suggested a few typo corrections / coding style suggestions.

**Comment by mkoeppe** [comment]
@tscrim do you want to make any of the suggested changes?

File: src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py (hunk 14 above)
Comment (tscrim) [reply src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py:315]: I find the spacing here is more readable considering the line above.

Comment (mkoeppe) [reply src/sage/algebras/hecke_algebras/ariki_koike_specht_modules.py:315]: ok

**Comment by mkoeppe** [comment]
Test failures, syntax error

**Comment by tscrim** [comment]
I remember testing this file...perhaps I used the wrong local copy. This should fix it now.

**Pull request merged**